File: ios/OpenEOSControl/App/CubeLut.swift
import CoreImage
import Foundation
import UIKit

struct CubeLut: Identifiable, Hashable, Sendable {
    let id = UUID()
    let name: String
    let size: Int
    let domainMin: [Float]
    let domainMax: [Float]
    let values: [Float]
    let rgbaData: Data

    init(name: String, size: Int, domainMin: [Float], domainMax: [Float], values: [Float]) {
        self.name = name
        self.size = size
        self.domainMin = domainMin
        self.domainMax = domainMax
        self.values = values
        var rgba = [Float]()
        rgba.reserveCapacity(size * size * size * 4)
        for index in stride(from: 0, to: values.count, by: 3) {
            rgba.append(values[index])
            rgba.append(values[index + 1])
            rgba.append(values[index + 2])
            rgba.append(1)
        }
        rgbaData = rgba.withUnsafeBytes { Data($0) }
    }

    static func == (lhs: CubeLut, rhs: CubeLut) -> Bool { lhs.id == rhs.id }

    func hash(into hasher: inout Hasher) { hasher.combine(id) }

    func sample(red: Float, green: Float, blue: Float) -> [Float] {
        let rc = coordinate(red, channel: 0)
        let gc = coordinate(green, channel: 1)
        let bc = coordinate(blue, channel: 2)
        let r0 = Int(floor(rc))
        let g0 = Int(floor(gc))
        let b0 = Int(floor(bc))
        let r1 = min(size - 1, r0 + 1)
        let g1 = min(size - 1, g0 + 1)
        let b1 = min(size - 1, b0 + 1)
        let rt = rc - Float(r0)
        let gt = gc - Float(g0)
        let bt = bc - Float(b0)
        return (0..<3).map { channel in
            let c00 = lerp(value(r0, g0, b0, channel), value(r1, g0, b0, channel), rt)
            let c10 = lerp(value(r0, g1, b0, channel), value(r1, g1, b0, channel), rt)
            let c01 = lerp(value(r0, g0, b1, channel), value(r1, g0, b1, channel), rt)
            let c11 = lerp(value(r0, g1, b1, channel), value(r1, g1, b1, channel), rt)
            return min(1, max(0, lerp(lerp(c00, c10, gt), lerp(c01, c11, gt), bt)))
        }
    }

    private func coordinate(_ value: Float, channel: Int) -> Float {
        min(1, max(0, (value - domainMin[channel]) / (domainMax[channel] - domainMin[channel])))
            * Float(size - 1)
    }

    private func value(_ red: Int, _ green: Int, _ blue: Int, _ channel: Int) -> Float {
        values[((blue * size * size + green * size + red) * 3) + channel]
    }
}

enum CubeLutError: LocalizedError {
    case invalid(String)

    var errorDescription: String? {
        switch self {
        case let .invalid(message): message
        }
    }
}

func parseCubeLut(_ text: String, fallbackName: String) throws -> CubeLut {
    guard text.utf8.count <= maximumCubeLutBytes else {
        throw CubeLutError.invalid("3D LUT exceeds the 16 MiB limit.")
    }
    var title: String?
    var size: Int?
    var domainMin: [Float] = [0, 0, 0]
    var domainMax: [Float] = [1, 1, 1]
    var hasDomainMin = false
    var hasDomainMax = false
    var values: [Float] = []

    for (zeroBasedLine, sourceLine) in text.split(separator: "\n", omittingEmptySubsequences: false).enumerated() {
        let line = String(sourceLine).split(separator: "#", maxSplits: 1, omittingEmptySubsequences: false)[0]
            .trimmingCharacters(in: .whitespacesAndNewlines)
        if line.isEmpty { continue }
        let tokens = line.split(whereSeparator: { $0.isWhitespace }).map(String.init)
        let lineNumber = zeroBasedLine + 1
        switch tokens[0].uppercased() {
        case "TITLE":
            guard title == nil else { throw CubeLutError.invalid("Duplicate TITLE at line \(lineNumber).") }
            let raw = String(line.dropFirst(tokens[0].count)).trimmingCharacters(in: .whitespaces)
            title = raw.trimmingCharacters(in: CharacterSet(charactersIn: "\"")).nilIfBlank
        case "LUT_3D_SIZE":
            guard size == nil, values.isEmpty, tokens.count == 2,
                  let parsed = Int(tokens[1]), (minimumCubeLutSize...maximumCubeLutSize).contains(parsed) else {
                throw CubeLutError.invalid("3D LUT size must be between 2 and 64 (line \(lineNumber)).")
            }
            size = parsed
            values.reserveCapacity(parsed * parsed * parsed * 3)
        case "DOMAIN_MIN":
            guard !hasDomainMin, values.isEmpty else {
                throw CubeLutError.invalid("Duplicate or late DOMAIN_MIN at line \(lineNumber).")
            }
            domainMin = try parseCubeVector(tokens, lineNumber: lineNumber)
            hasDomainMin = true
        case "DOMAIN_MAX":
            guard !hasDomainMax, values.isEmpty else {
                throw CubeLutError.invalid("Duplicate or late DOMAIN_MAX at line \(lineNumber).")
            }
            domainMax = try parseCubeVector(tokens, lineNumber: lineNumber)
            hasDomainMax = true
        case "LUT_3D_INPUT_RANGE":
            guard !hasDomainMin, !hasDomainMax, values.isEmpty, tokens.count == 3 else {
                throw CubeLutError.invalid("Duplicate or conflicting input range at line \(lineNumber).")
            }
            let minimum = try finiteFloat(tokens[1], lineNumber: lineNumber)
            let maximum = try finiteFloat(tokens[2], lineNumber: lineNumber)
            domainMin = [minimum, minimum, minimum]
            domainMax = [maximum, maximum, maximum]
            hasDomainMin = true
            hasDomainMax = true
        case "LUT_1D_SIZE":
            throw CubeLutError.invalid("1D and shaper LUTs are not supported.")
        default:
            guard let size else {
                throw CubeLutError.invalid("LUT_3D_SIZE must appear before table data (line \(lineNumber)).")
            }
            guard tokens.count == 3 else { throw CubeLutError.invalid("Invalid 3D LUT row at line \(lineNumber).") }
            values.append(contentsOf: try tokens.map { try finiteFloat($0, lineNumber: lineNumber) })
            guard values.count <= size * size * size * 3 else {
                throw CubeLutError.invalid("3D LUT contains more rows than LUT_3D_SIZE declares.")
            }
        }
    }

    guard let size else { throw CubeLutError.invalid("LUT_3D_SIZE is required.") }
    guard zip(domainMin, domainMax).allSatisfy({ $0.isFinite && $1.isFinite && $0 < $1 }) else {
        throw CubeLutError.invalid("Every LUT domain minimum must be lower than its maximum.")
    }
    let expectedRows = size * size * size
    guard values.count == expectedRows * 3 else {
        throw CubeLutError.invalid("3D LUT requires \(expectedRows) RGB rows; found \(values.count / 3).")
    }
    let fallback = URL(fileURLWithPath: fallbackName).deletingPathExtension().lastPathComponent
    let name = String((title ?? fallback.nilIfBlank ?? "Imported LUT").prefix(maximumCubeLutNameLength))
    return CubeLut(name: name, size: size, domainMin: domainMin, domainMax: domainMax, values: values)
}

func renderCubeLutPreview(data: Data, lut: CubeLut) -> UIImage? {
    guard let input = CIImage(data: data, options: [.applyOrientationProperty: true]) else { return nil }
    let normalization = CIFilter(name: "CIColorMatrix")
    normalization?.setValue(input, forKey: kCIInputImageKey)
    let inverse = zip(lut.domainMin, lut.domainMax).map { 1 / ($1 - $0) }
    normalization?.setValue(CIVector(x: CGFloat(inverse[0]), y: 0, z: 0, w: 0), forKey: "inputRVector")
    normalization?.setValue(CIVector(x: 0, y: CGFloat(inverse[1]), z: 0, w: 0), forKey: "inputGVector")
    normalization?.setValue(CIVector(x: 0, y: 0, z: CGFloat(inverse[2]), w: 0), forKey: "inputBVector")
    normalization?.setValue(CIVector(x: 0, y: 0, z: 0, w: 1), forKey: "inputAVector")
    normalization?.setValue(
        CIVector(
            x: CGFloat(-lut.domainMin[0] * inverse[0]),
            y: CGFloat(-lut.domainMin[1] * inverse[1]),
            z: CGFloat(-lut.domainMin[2] * inverse[2]),
            w: 0
        ),
        forKey: "inputBiasVector"
    )
    guard let normalized = normalization?.outputImage else { return nil }
    let clamp = CIFilter(name: "CIColorClamp")
    clamp?.setValue(normalized, forKey: kCIInputImageKey)
    clamp?.setValue(CIVector(x: 0, y: 0, z: 0, w: 0), forKey: "inputMinComponents")
    clamp?.setValue(CIVector(x: 1, y: 1, z: 1, w: 1), forKey: "inputMaxComponents")
    guard let clamped = clamp?.outputImage else { return nil }
    let colorCube = CIFilter(name: "CIColorCube")
    colorCube?.setValue(clamped, forKey: kCIInputImageKey)
    colorCube?.setValue(lut.size, forKey: "inputCubeDimension")
    colorCube?.setValue(lut.rgbaData, forKey: "inputCubeData")
    guard let output = colorCube?.outputImage,
          let cgImage = cubeLutContext.value.createCGImage(output, from: input.extent) else { return nil }
    return UIImage(cgImage: cgImage)
}

let maximumCubeLutBytes = 16 * 1024 * 1024

private func parseCubeVector(_ tokens: [String], lineNumber: Int) throws -> [Float] {
    guard tokens.count == 4 else { throw CubeLutError.invalid("Invalid LUT domain at line \(lineNumber).") }
    return try tokens.dropFirst().map { try finiteFloat($0, lineNumber: lineNumber) }
}

private func finiteFloat(_ value: String, lineNumber: Int) throws -> Float {
    guard let number = Float(value), number.isFinite else {
        throw CubeLutError.invalid("Invalid finite number at line \(lineNumber).")
    }
    return number
}

private func lerp(_ start: Float, _ end: Float, _ amount: Float) -> Float { start + (end - start) * amount }

private extension String {
    var nilIfBlank: String? { trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? nil : self }
}

private final class SendableCIContext: @unchecked Sendable {
    let value = CIContext(options: [.cacheIntermediates: false])
}

private let cubeLutContext = SendableCIContext()
private let minimumCubeLutSize = 2
private let maximumCubeLutSize = 64
private let maximumCubeLutNameLength = 120

File: android/app/src/main/java/dev/openeos/control/ui/CubeLut.kt
package dev.openeos.control.ui

import android.graphics.Bitmap
import kotlin.math.floor
import kotlin.math.roundToInt

data class CubeLut(
    val name: String,
    val size: Int,
    val domainMin: FloatArray,
    val domainMax: FloatArray,
    val values: FloatArray,
) {
    init {
        require(name.isNotBlank())
        require(size in MIN_CUBE_LUT_SIZE..MAX_CUBE_LUT_SIZE)
        require(domainMin.size == 3 && domainMax.size == 3)
        require(values.size == size * size * size * 3)
    }

    fun sample(red: Float, green: Float, blue: Float): FloatArray {
        val redCoordinate = coordinate(red, 0)
        val greenCoordinate = coordinate(green, 1)
        val blueCoordinate = coordinate(blue, 2)
        val r0 = floor(redCoordinate).toInt()
        val g0 = floor(greenCoordinate).toInt()
        val b0 = floor(blueCoordinate).toInt()
        val r1 = (r0 + 1).coerceAtMost(size - 1)
        val g1 = (g0 + 1).coerceAtMost(size - 1)
        val b1 = (b0 + 1).coerceAtMost(size - 1)
        val rt = redCoordinate - r0
        val gt = greenCoordinate - g0
        val bt = blueCoordinate - b0
        return FloatArray(3) { channel -> sampleChannel(r0, g0, b0, r1, g1, b1, rt, gt, bt, channel) }
    }

    private fun coordinate(value: Float, channel: Int): Float =
        ((value - domainMin[channel]) / (domainMax[channel] - domainMin[channel]))
            .coerceIn(0f, 1f) * (size - 1)

    private fun value(red: Int, green: Int, blue: Int, channel: Int): Float =
        values[((blue * size * size + green * size + red) * 3) + channel]

    private fun sampleChannel(
        r0: Int,
        g0: Int,
        b0: Int,
        r1: Int,
        g1: Int,
        b1: Int,
        rt: Float,
        gt: Float,
        bt: Float,
        channel: Int,
    ): Float {
        val c00 = lerp(value(r0, g0, b0, channel), value(r1, g0, b0, channel), rt)
        val c10 = lerp(value(r0, g1, b0, channel), value(r1, g1, b0, channel), rt)
        val c01 = lerp(value(r0, g0, b1, channel), value(r1, g0, b1, channel), rt)
        val c11 = lerp(value(r0, g1, b1, channel), value(r1, g1, b1, channel), rt)
        return lerp(lerp(c00, c10, gt), lerp(c01, c11, gt), bt).coerceIn(0f, 1f)
    }

    fun sampleArgb(pixel: Int): Int {
        val redCoordinate = coordinate((pixel shr 16 and 0xff) / 255f, 0)
        val greenCoordinate = coordinate((pixel shr 8 and 0xff) / 255f, 1)
        val blueCoordinate = coordinate((pixel and 0xff) / 255f, 2)
        val r0 = floor(redCoordinate).toInt()
        val g0 = floor(greenCoordinate).toInt()
        val b0 = floor(blueCoordinate).toInt()
        val r1 = (r0 + 1).coerceAtMost(size - 1)
        val g1 = (g0 + 1).coerceAtMost(size - 1)
        val b1 = (b0 + 1).coerceAtMost(size - 1)
        val rt = redCoordinate - r0
        val gt = greenCoordinate - g0
        val bt = blueCoordinate - b0
        val red = sampleChannel(r0, g0, b0, r1, g1, b1, rt, gt, bt, 0)
        val green = sampleChannel(r0, g0, b0, r1, g1, b1, rt, gt, bt, 1)
        val blue = sampleChannel(r0, g0, b0, r1, g1, b1, rt, gt, bt, 2)
        return (pixel and 0xff000000.toInt()) or
            ((red * 255f).roundToInt().coerceIn(0, 255) shl 16) or
            ((green * 255f).roundToInt().coerceIn(0, 255) shl 8) or
            (blue * 255f).roundToInt().coerceIn(0, 255)
    }
}

internal fun parseCubeLut(text: String, fallbackName: String): CubeLut {
    require(text.toByteArray(Charsets.UTF_8).size <= MAX_CUBE_LUT_BYTES) {
        "3D LUT exceeds the ${MAX_CUBE_LUT_BYTES / (1024 * 1024)} MiB limit."
    }
    var title: String? = null
    var size: Int? = null
    var domainMin = floatArrayOf(0f, 0f, 0f)
    var domainMax = floatArrayOf(1f, 1f, 1f)
    var hasDomainMin = false
    var hasDomainMax = false
    val values = ArrayList<Float>()

    text.lineSequence().forEachIndexed { index, sourceLine ->
        val line = sourceLine.substringBefore('#').trim()
        if (line.isEmpty()) return@forEachIndexed
        val tokens = line.split(Regex("\\s+"))
        when (tokens.first().uppercase()) {
            "TITLE" -> {
                require(title == null) { "Duplicate TITLE at line ${index + 1}." }
                title = line.substringAfter(' ', "").trim().trim('"').takeIf(String::isNotBlank)
            }
            "LUT_3D_SIZE" -> {
                require(size == null && values.isEmpty()) { "Duplicate or late LUT_3D_SIZE at line ${index + 1}." }
                require(tokens.size == 2) { "Invalid LUT_3D_SIZE at line ${index + 1}." }
                size = tokens[1].toIntOrNull()?.takeIf { it in MIN_CUBE_LUT_SIZE..MAX_CUBE_LUT_SIZE }
                    ?: error("3D LUT size must be between $MIN_CUBE_LUT_SIZE and $MAX_CUBE_LUT_SIZE.")
            }
            "DOMAIN_MIN" -> {
                require(!hasDomainMin && values.isEmpty()) { "Duplicate or late DOMAIN_MIN at line ${index + 1}." }
                domainMin = parseVector(tokens, index)
                hasDomainMin = true
            }
            "DOMAIN_MAX" -> {
                require(!hasDomainMax && values.isEmpty()) { "Duplicate or late DOMAIN_MAX at line ${index + 1}." }
                domainMax = parseVector(tokens, index)
                hasDomainMax = true
            }
            "LUT_3D_INPUT_RANGE" -> {
                require(!hasDomainMin && !hasDomainMax && values.isEmpty()) {
                    "Duplicate or conflicting input range at line ${index + 1}."
                }
                require(tokens.size == 3) { "Invalid LUT_3D_INPUT_RANGE at line ${index + 1}." }
                val minimum = tokens[1].toFiniteFloat(index)
                val maximum = tokens[2].toFiniteFloat(index)
                domainMin = FloatArray(3) { minimum }
                domainMax = FloatArray(3) { maximum }
                hasDomainMin = true
                hasDomainMax = true
            }
            "LUT_1D_SIZE" -> error("1D and shaper LUTs are not supported.")
            else -> {
                require(size != null) { "LUT_3D_SIZE must appear before table data (line ${index + 1})." }
                require(tokens.size == 3) { "Invalid 3D LUT row at line ${index + 1}." }
                tokens.forEach { values += it.toFiniteFloat(index) }
                require(values.size <= size!! * size!! * size!! * 3) {
                    "3D LUT contains more rows than LUT_3D_SIZE declares."
                }
            }
        }
    }

    val cubeSize = requireNotNull(size) { "LUT_3D_SIZE is required." }
    require(domainMin.indices.all { domainMin[it].isFinite() && domainMax[it].isFinite() && domainMin[it] < domainMax[it] }) {
        "Every LUT domain minimum must be lower than its maximum."
    }
    require(values.size == cubeSize * cubeSize * cubeSize * 3) {
        "3D LUT requires ${cubeSize * cubeSize * cubeSize} RGB rows; found ${values.size / 3}."
    }
    val safeName = title?.take(MAX_CUBE_LUT_NAME_LENGTH)
        ?: fallbackName.substringBeforeLast('.').trim().take(MAX_CUBE_LUT_NAME_LENGTH).ifBlank { "Imported LUT" }
    return CubeLut(safeName, cubeSize, domainMin, domainMax, values.toFloatArray())
}

internal fun applyCubeLut(
    bitmap: Bitmap,
    lut: CubeLut,
    checkCancellation: () -> Unit = {},
): Bitmap {
    require(bitmap.width > 0 && bitmap.height > 0)
    val pixels = IntArray(bitmap.width * bitmap.height)
    bitmap.getPixels(pixels, 0, bitmap.width, 0, 0, bitmap.width, bitmap.height)
    pixels.indices.forEach { index ->
        if (index % (bitmap.width * 8).coerceAtLeast(1) == 0) checkCancellation()
        val pixel = pixels[index]
        pixels[index] = lut.sampleArgb(pixel)
    }
    val result = Bitmap.createBitmap(pixels, bitmap.width, bitmap.height, Bitmap.Config.ARGB_8888)
    return try {
        checkCancellation()
        result
    } catch (error: Throwable) {
        result.recycle()
        throw error
    }
}

private fun parseVector(tokens: List<String>, lineIndex: Int): FloatArray {
    require(tokens.size == 4) { "Invalid LUT domain at line ${lineIndex + 1}." }
    return FloatArray(3) { tokens[it + 1].toFiniteFloat(lineIndex) }
}

private fun String.toFiniteFloat(lineIndex: Int): Float =
    toFloatOrNull()?.takeIf(Float::isFinite) ?: error("Invalid finite number at line ${lineIndex + 1}.")

private fun lerp(start: Float, end: Float, amount: Float): Float = start + (end - start) * amount

internal const val MAX_CUBE_LUT_BYTES = 16 * 1024 * 1024
private const val MIN_CUBE_LUT_SIZE = 2
private const val MAX_CUBE_LUT_SIZE = 64
private const val MAX_CUBE_LUT_NAME_LENGTH = 120
